contracts/apps/TokenIOU.sol

```diff
@@ -0,0 +1,214 @@
+    mapping(bytes32 redeemHash => uint48 time) public redeemed;
+
+    // Redeem event
+    event Redeem(address indexed from, address indexed to, uint256 amount);
+
+    function initialize(IERC20Upgradeable _token) public initializer {
+        __Ownable_init();
+        __UUPSUpgradeable_init();
+        token = _token;
+    }
+
+    /**
+        * @notice Get the hash of the redeem request
+        * @param from       The address of the redeemer
+        * @param amount     The amount of tokens to redeem
+        * @param validUntil The timestamp until which the redeem is valid
+        * @param validAfter  The timestamp from which the redeem is valid
+        * @param nonce      A unique number to prevent replay attacks
+        * @return The hash of the redeem request
+
+        Can be used to easily generate a valid hash for the redeem function
+    */
+    function getHash(
+        address from,
+        uint256 amount,
+        uint48 validUntil,
+        uint48 validAfter,
+        uint256 nonce
+    ) public view returns (bytes32) {
+        return
+            keccak256(
+                abi.encodePacked(
+                    from,
+                    amount,
+                    validUntil,
+                    validAfter,
+                    nonce,
+                    block.chainid,
+                    address(this)
+                )
+            );
+    }
+
+    /**
+     * @notice Redeem tokens
+     * @param from       The address of the redeemer
+     * @param amount     The amount of tokens to redeem
+     * @param validUntil The timestamp until which the redeem is valid
+     * @param validAfter  The timestamp from which the redeem is valid
+     * @param nonce      A unique number to prevent replay attacks
+     * @param signature  The signature of the redeem request
+     */
+    function redeem(
+        address from,
+        uint256 amount,
+        uint48 validUntil,
+        uint48 validAfter,
+        uint256 nonce,
+        bytes calldata signature
+    ) public {
+        uint48 currentTime = uint48(block.timestamp);
+        require(currentTime >= validAfter, "TokenIOU: expired or not due");
+        require(currentTime < validUntil, "TokenIOU: expired or not due");
+
+        bytes32 redeemHash = keccak256(
+            abi.encodePacked(
+                "\x19Ethereum Signed Message:\n32",
+                getHash(from, amount, validUntil, validAfter, nonce)
+            )
+        );
+
+        require(redeemed[redeemHash] == 0, "TokenIOU: Already redeemed");
+
+        // check the sender's signature
+        address to = msg.sender;
+
+        // check if the msg.sender is a smart contract
+        if (_contractExists(to)) {
+            // check ownership of the contract
+            IERC1271 account = IERC1271(to);
+
+            require(
+                account.isValidSignature(redeemHash, signature) == 0x1626ba7e,
+                "TokenIOU: Invalid signature"
```

Comment:
TokenIOU is ugly (also not needed in those error messages). Simply "Invalid signature" will do.

Maybe a better name than TokenIOU is IOUEntryPoint
Or IOUJesus (the OG redeemer)

Reply:
It's better to specify a prefix, because this will be called from contracts and we don't want to confuse the error with another contract error. 

see how open zeppelin does it: `${contractName}: ${errorString}`
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/bd325d56b4c62c9c5c1aff048c37c6bb18ac0290/contracts/token/ERC20/ERC20.sol#L309

Reply:
Could've just kept `IOU` as a contract name as well. But it's specific for ERC20. So, `ERC20IOU`?